### ISSUE TITLE
CA1859 has false negative for all types having shared base type

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn-analyzers#6415]
### Analyzer

**Diagnostic ID**: CA1859

### Analyzer source

https://github.com/dotnet/roslyn-analyzers/commit/6ae2683c5033dfde0b73d4e73b372f00ca6ceacd

### Describe the bug

If a method is declared as returning type A, and all of the returns are for types that are or derive from B which derives from A, I'd expect the analyzer to warn that the method should instead be declared with B as a return type, but there's no such warning issued.

### Steps To Reproduce

```C#
[Fact]
public static async Task Repro()
{
    const string Source = @"
        #nullable enable

        using System;
        using System.IO;

        internal class C
        {
            private Stream GetStream(int i)
            {
                if (i == 0) return new MyStream1();
                return new MyStream2();
            }
        }

        class MyStream1 : MemoryStream { }
        class MyStream2 : MemoryStream { }
        ";

    await TestCSAsync(Source);
}
```

### Expected behavior

A warning saying that GetStream should return MemoryStream rather than Stream.

### Actual behavior

No warning.

cc: @geeknoid, @buyaa-n

**Comment by geeknoid** [comment]
Note that in general, the analyzer only triggers if changing the type will result in a reduction in the number of virtual and/or interface calls. If you don't make a virtual call on a given instance, there's no point in changing its type. This is likely what's happening here.

**Comment by stephentoub** [comment]
I just added:
```C#
                    public int Read()
                    {
                        return GetStream(0).ReadByte();
                    }
```
to the repro, and it still doesn't trigger the analyzer.

**Comment by stephentoub** [comment]
Also, I just tried:
```C#
        [Fact]
        public static async Task Repro()
        {
            const string Source = @"
                #nullable enable

                using System.Threading.Tasks;

                internal class C
                {
                    private Task Get()
                    {
                        return Task.FromResult(42);
                    }
                }
                ";

            await TestCSAsync(Source);
        }
```
and that _does_ issue a warning, even though no one's using the method at all and thus isn't making any virtual calls to the returned instance.

Have I misunderstood then?

**Comment by geeknoid** [comment]
OK, sorry, I read your first example again. The analyzer currently doesn't contain any logic to do what you're asking. As a general rule, it triggers iff all assignments to the value are of a single more derived type, which when used would eliminate virtual dispatches. So in this case, the function is returning two different types, so this won't be flagged. In addition, even if the functionality was provided to settle on some common base class, the example would still not trigger because the common base class of MemoryStream would not reduce the number of virtual calls made (since ReadByte is still virtual on MemoryStream)

**Comment by geeknoid** [comment]
Regarding the second case with getting a task, the check to ensure a change results in a reduction of virtual calls only applies for local variables, fields, and parameters. But for method return types, the check isn't applied there as method returns can have a transitive effect I can't readily analyze. Basically, if you upgrade a method return, you can make it possible to upgrade a local variable somewhere, which is turn can eliminate a virtual call.

So we end up in that case with a bit of 'peeling the onion'. You'll get a message telling you you can upgrade the return type of a method. Once you rebuild, you'll get a new warning telling you to upgrade some local variable to which you assign the result of the method call. This can in turn result in another warning telling you to upgrade the second method's return type, etc.

Doing a full analysis to provide a one-shot capture of all possible upgrades would be considerably more involved than the analyzer's logic currently allows for.

**Comment by geeknoid** [comment]
So I'm inclined to close this issue as Won't Fix since things are working as expected.
